### PR TITLE
[stable/mongodb] Statefulset standalone

### PR DESCRIPTION
**Pull request by dargolith**
#### What this PR does / why we need it:
This PR makes it possible to set a variable ("useStatefulSet") in the chart values to true in order to use StatefulSet instead of Deployment even when using replicaSet.enabled=false. StatefulSet is good to use in order to prevent automatic removal of pvcs when deleting releases. It also makes it possible to re-attach to pvcs (volumes) if redeploying the same release.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by carrodher** [comment]
Hi, a few comments:
- There is a `values-production.yaml` where you should also implement the changes
- What about the file name? It is called `deployment-standalone.yaml` although it's possible that it was deployed as a statefulset.

What do you think about those changes @juan131?

**Comment by dargolith** [comment]
@carrodher Thanks for feedback!

> * There is a `values-production.yaml` where you should also implement the changes
True! I will add that. I guess the default value false makes most sense there too since it won't even be used for replicaSets.enabled=true.

> * What about the file name? It is called `deployment-standalone.yaml` although it's possible that it was deployed as a statefulset.
I pondered that one as well but came to the conclusion that deployment was the more generic word, even though it also is the name of the deployment kind. Deployment is also the default-value. Another suggestion could be just "standalone.yaml" if that isn't too cryptic.

**Comment by juan131** [comment]
Hi everyone,

Is there any other reason (apart of preventing Helm to remove the PVCs when removing a release) to use a **statefulset** instead of a **deployment**? 

The reasons I know to use statefulsets in favor of deployments are attatched to the asset requirements. For instance, you need to use statefulsets when the asset has one of the following requirements:

- Stable, unique network identifiers.
- Stable, unique persistent storage per pod.
- Ordered, graceful deployment and scaling.
- Ordered, graceful deletion and termination
- Ordered, automated rolling updates.

I don't think that a "standalone" PostgreSQL setup meets any of those requirements.

On the other hand, this chart allows you to select the PVC to use (via the `persistence.existingClaim` parameter). So you can:

- Create a PVC
- Install a chart using the `persistence.existingClaim`  parameter and use the created PVC
- Remove the chart (the PVC won't be removed)
- Reuse the PVC on other releases.

**Comment by dargolith** [comment]
The main reason is not only to prevent deletion, it is that the statefulset allows re-attachment of the volume on redeployment. This makes the handling of the pvc easier since it will be created if non-existent and re-attach if it exists. I also don't have to make my own pvc, though the pvc is kind of simple I guess, but still it feels like it should be included in the required chart rather.

> I don't think that a "standalone" PostgreSQL setup meets any of those requirements.

@juan131 You mean mongodb, right?

I've read different discussions on both github / stackoverflow etc where this kind of behavior is requested and in the end statefulset is being recommended. But also some other guidelines from kubernetes / google that I think suggest this:

https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes#deployments_vs_statefulsets

https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#using-statefulsets

If there's problems with allowing this feature and if there's another way to achieve this in the same "automated" way I'd like to hear about that suggestion. I'd be fine to use it if so. My research so far has only led to suggestions on using statefulset since it is recommended for that kind of persistence. As long as I can define my chart with mongodb as requirement and create those automated features:

* My own chart starting from install without preparation, creating a pvc if it doesn't exist
* helm delete --purge without removing the pvc and lose data
* helm install again after helm delete and get the same pvc reattached for the db if the pvc exists

**Comment by vsliouniaev** [comment]
Here is the guidance from this repo:

https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md#kubernetes-native-workloads
> Workloads that are stateful in nature such as databases, key-value stores, message queues, in-memory caches are to be created as StatefulSets

File: stable/mongodb/README.md [hunk 1]
```diff
@@ -122,6 +122,7 @@ The following table lists the configurable parameters of the MongoDB chart and t
 | `persistence.size`                                 | Size of data volume                                                                                                                                       | `8Gi`                                                    |
 | `persistence.annotations`                          | Persistent Volume annotations                                                                                                                             | `{}`                                                     |
 | `persistence.existingClaim`                        | Name of an existing PVC to use (avoids creating one if this is given)                                                                                     | `nil`                                                    |
+| `useStatefulSet`                                   | Set to true to use StatefulSet instead of Deployment even when replicaSet.enalbed=false                                                                   | `nil`                                                    |
```
Comment (juan131) [inline stable/mongodb/README.md:132]: Typo: `enalbed` -> `enabled`

**Review by juan131 — APPROVED** [review]

**Comment by juan131** [comment]
Hi @dargolith 

Sorry for the late response, I lost the track of this PR.

I guess it makes no harm to support both setups.

> @juan131 You mean mongodb, right?

Yes I meant MongoDB.

**Comment by juan131** [comment]
Could you rebase your PR from master and solve the conflicts?

**Comment by dargolith** [comment]
@juan131 Hi, sorry from me as well. Been lots of stuff over christmas and I somehow missed this.

I'll do the rebase!

**Comment by dargolith** [comment]
@juan131 Erm, did I do something wrong during the rebase? It looks kind of odd...

**Comment by juan131** [comment]
Hi @dargolith 

Sorry for the late response. I don't think anything odd in the rebase but it look like there are still conflicts with the current master branch. 

Could you address them? sorry for the inconveniences.

**Comment by dargolith** [comment]
@juan131 Hi! I did another rebase. :)

**Review by juan131 — APPROVED** [review]
/ok-to-test
/lgtm

**Comment by juan131** [comment]
Hi @dargolith 

Thanks so much for your contribution!!! Your PR was merged

**Comment by dargolith** [comment]
@juan131 Glad to contribute, thank you for a great chart! :)

**Pull request merged**